### DISCUSSION TITLE
Seeking Best Practices for Clean Integration of shadcn UI Theming with Radix UI Colors and Tailwind CSS

### DISCUSSION THREAD

**Discussion by nuluh** [shadcn-ui/ui#3413] [General]
Hello everyone! I’m currently working on a project where I’m utilizing shadcn UI theming and have recently discovered Radix UI colors. Both systems offer fantastic features, and I’m keen to incorporate them together in my project, along with Tailwind CSS.

While Tailwind CSS provides a built-in color palette, I found that it doesn’t entirely suit my project’s needs. Therefore, I’m interested in **combining shadcn UI and Radix UI colors** to create a more tailored palette.

I was wondering if anyone could share their best practices for integrating these systems, with a **focus on clean architecture, efficient workflow, semantic code, and readability while keeping the implementation easy to maintain**.

Some specific areas I’m interested in include managing shadcn/ui color CSS variables alongside Radix UI Theme and potential strategies to avoid conflicts between the two systems, as well as insights on how to effectively use Tailwind CSS to create a custom color palette.

Additionally, I’m facing challenges choosing the right colors from Radix UI, as explained in their documentation on understanding the color scale (https://www.radix-ui.com/colors/docs/palette-composition/understanding-the-scale). 

Also, when I change the color behaviours of a button with my own Tailwind CSS utility classes, what happen to the component if there's an update from shadcn itself? Will my edited button components styling gone?

I’d appreciate any advice on selecting colors for particular elements and applying them using Tailwind CSS classes.

Any advice, practical tips, or experiences you’ve had with this would be greatly appreciated.

Thank you in advance for your insights, and I look forward to learning from your experiences.

**Comment by ewgenius** [comment]
Up on this, I'm also currently experimenting on how to integrate radix colors with shadcn ui.

Before starting to use shadcn components, I've made a custom tailwind plugin, to expand radix colors into tailwind scales through env variables automatically:
https://tailwindcss-variable-themes.vercel.app/radix-semantic
https://github.com/ewgenius/tailwindcss-variable-themes

But that won't fit well into shadcn styles unfortunately, so still don't have an ideal solution.

**Comment by jdhenry08** [comment]
Same here. I just came across this, which might help:
[https://github.com/wallpants/hsl-radix-colors](https://github.com/wallpants/hsl-radix-colors)

**Comment by ewgenius** [comment]
I've just made radix theme generator for shadcn/ui https://ui.ewgenius.me/shadcn-radix-colors - that will generate a css theme replacement.
Also planning to publish some components modifications, which will use [radix colors scale](https://www.radix-ui.com/colors/docs/palette-composition/understanding-the-scale) fully.

**Comment by leonardbiofi** [comment]
Sam here. Whenever you find something that works well ! There are couple of examples in which Radix-colors are integrated with tailwindcss. With tailwind v4 should be easier to achieve though !
